x/evm/types/statedb.go

```diff
@@ -685,27 +693,29 @@ func (csdb *CommitStateDB) Copy() *CommitStateDB {
 		ctx:                  csdb.ctx,
 		storeKey:             csdb.storeKey,
 		accountKeeper:        csdb.accountKeeper,
-		stateObjects:         make([]stateEntry, len(csdb.journal.dirties)),
-		addressToObjectIndex: make(map[ethcmn.Address]int, len(csdb.journal.dirties)),
-		stateObjectsDirty:    make(map[ethcmn.Address]struct{}, len(csdb.journal.dirties)),
+		stateObjects:         []stateEntry{},
+		addressToObjectIndex: make(map[ethcmn.Address]int),
+		stateObjectsDirty:    make(map[ethcmn.Address]struct{}),
 		refund:               csdb.refund,
 		logSize:              csdb.logSize,
-		preimages:            make(map[ethcmn.Hash][]byte),
+		preimages:            make([]preimageEntry, len(csdb.preimages)),
+		hashToPreimageIndex:  make(map[ethcmn.Hash]int, len(csdb.hashToPreimageIndex)),
 		journal:              newJournal(),
 	}
 
 	// copy the dirty states, logs, and preimages
-	for i, dirty := range csdb.journal.dirties {
+	for _, dirty := range csdb.journal.dirties {
 		// There is a case where an object is in the journal but not in the
 		// stateObjects: OOG after touch on ripeMD prior to Byzantium. Thus, we
 		// need to check for nil.
 		//
 		// Ref: https://github.com/ethereum/go-ethereum/pull/16485#issuecomment-380438527
 		if idx, exist := csdb.addressToObjectIndex[dirty.address]; exist {
-			state.stateObjects[i] = stateEntry{
+			state.stateObjects = append(state.stateObjects, stateEntry{
 				address:     dirty.address,
 				stateObject: csdb.stateObjects[idx].stateObject.deepCopy(state),
-			}
+			})
+			state.addressToObjectIndex[dirty.address] = len(state.stateObjects) - 1
```

Comment:
the `addressToObjectIndex` was not being updated either

Reply:
not sure why append should reduce by 1.